paddle/framework/data_transform.cc

```diff
@@ -23,5 +24,68 @@ DataTransformFnMap& DataTransformFnMap::Instance() {
   return data_transform_map;
 }
 
+auto kernel_FP32 = OpKernelType(proto::DataType::FP32, platform::CPUPlace(),
+                                DataLayout::kNHWC, LibraryType::kPlain);
+
+auto kernel_FP64 = OpKernelType(proto::DataType::FP64, platform::CPUPlace(),
+                                DataLayout::kNHWC, LibraryType::kPlain);
+
+auto kernel_Layout0 = OpKernelType(proto::DataType::FP64, platform::CPUPlace(),
+                                   DataLayout::kNHWC, LibraryType::kPlain);
+
+auto kernel_Layout1 = OpKernelType(proto::DataType::FP64, platform::CPUPlace(),
+                                   DataLayout::kNCHW, LibraryType::kPlain);
+
+void TransDataType(const platform::DeviceContext* ctx,
+                   const KernelTypePair& kernel_pair, const Variable& in,
+                   Variable* out) {
+  PADDLE_ENFORCE(in.IsType<Tensor>(), "Only Support Tensor transform!.");
+  auto src = in.Get<Tensor>();
+  auto* dst = out->GetMutable<Tensor>();
+  // TODO(dzhwinter): CPU <-> GPU need a copy here
+
+  auto dims = src.dims();
+  dst->Resize(dims);
+  auto dst_type = kernel_pair.second.data_type_;
+  auto src_type = kernel_pair.first.data_type_;
+
+  switch (src_type) {
+    case proto::DataType::FP32:
+      framework::VisitDataType(dst_type, CastDataType<float>(src, dst, ctx));
+      break;
+    case proto::DataType::FP64:
+      framework::VisitDataType(dst_type, CastDataType<double>(src, dst, ctx));
+      break;
+    case proto::DataType::INT32:
+      framework::VisitDataType(dst_type, CastDataType<int>(src, dst, ctx));
+      break;
+    case proto::DataType::INT64:
+      framework::VisitDataType(dst_type, CastDataType<int64_t>(src, dst, ctx));
+      break;
+    case proto::DataType::BOOL:
+      framework::VisitDataType(dst_type, CastDataType<bool>(src, dst, ctx));
+      break;
+    default:
+      PADDLE_THROW("Not support type %d", src_type);
+  }
+}
+
+void TransDataLayout(const platform::DeviceContext* ctx,
+                     const KernelTypePair& kernel_pair, const Variable& in,
+                     Variable* out) {
+  PADDLE_ENFORCE(in.IsType<Tensor>(), "Only Support LoDTensor transform!.");
+  auto src = in.Get<Tensor>();
+  auto* dst = out->GetMutable<Tensor>();
+  dst->Resize(src.dims());
+  auto place = kernel_pair.second.place_;
+  CopyFrom(src, place, *ctx, dst);
```

Comment:
Maybe `TransDataLayout` is wrong.
e.g.  `arr`'s layout is `NCHW` format, and data is:
```
    [[[[ 0,  1,  2],
         [ 3,  4,  5]],

        [[ 6,  7,  8],
         [ 9, 10, 11]]]]
```
Obviously, `arr`'s shape is (1,2,2,3).
If we change `NCHW` to `NHWC`, it's shape will be (1,3,2,2), and the data is:
```
    [[[[ 0,  6],
         [ 3,  9]],

        [[ 1,  7],
         [ 4, 10]],

        [[ 2,  8],
         [ 5, 11]]]]
```

Reply:
Done.